File: quizzes8/06-problem.ts
export {};

// @ts-ignore
import * as fs from "fs";

type Feedback = {
  comment: string;
  date: string;
};

type Result = {
  rating: number;
  feedback: Feedback[];
};

type Staff = {
  id: number;
  full_name: string;
  email: string;
  result: Result;
};

type TopRatedStaff = Omit<Staff, "result"> & {
  rating: Result["rating"];
};

const staffs: Staff[] = JSON.parse(
  fs.readFileSync("input_data/staff.json", "utf8"),
);

const findTopRatedStaffsWithChained = (): TopRatedStaff[] => {
  return staffs.flatMap(({ result, ...staff }) => {
    if (result.rating <= 4) return [];

    return {
      ...staff,
      rating: result.rating,
    };
  });
};

const findTopRatedStaffsWithLoop = (): TopRatedStaff[] => {
  const topRatedStaffs: TopRatedStaff[] = [];

  for (const { result, ...staff } of staffs) {
    if (result.rating > 4) {
      topRatedStaffs.push({
        ...staff,
        rating: result.rating,
      });
    }
  }

  return topRatedStaffs;
};

console.log(JSON.stringify({ data: findTopRatedStaffsWithChained() }, null, 2));
console.log(JSON.stringify({ data: findTopRatedStaffsWithLoop() }, null, 2));

File: bins2/06-problem.rb
# frozen_string_literal: true

require 'json'

staffs = JSON.parse(
  File.read('input_data/staff.json'),
  symbolize_names: true
)

def find_top_rated_staffs_with_chain(staffs)
  staffs.filter_map do |staff|
    rating = staff[:result][:rating]
    next if rating <= 4

    staff.except(:result).merge(rating: rating)
  end
end

def find_top_rated_staffs_with_loop(staffs)
  top_rated = []

  staffs.each do |staff|
    rating = staff[:result][:rating]
    if rating > 4
      top_rated << staff.except(:result).merge(rating: rating)
    end
  end

  top_rated
end

puts JSON.pretty_generate({ data: find_top_rated_staffs_with_chain(staffs) })
puts JSON.pretty_generate({ data: find_top_rated_staffs_with_loop(staffs) })
